File: Haskell/22-knapsack.hs
import Control.Monad (replicateM)
import Data.Array (Array, listArray, range, (!))
import Data.List (sort)

data Item =  Item Int Int


convertToIntList :: String -> [Int]
convertToIntList = map read . words


readItems :: Int -> IO [Item]
readItems nrItems = replicateM nrItems $ fmap ((\[w, v] -> Item w v) . convertToIntList) getLine


findItemIndices :: Array (Int, Int) Int -> Array Int Item -> Int -> Int -> [Int]
findItemIndices knapsack items n capacity = go [] capacity n
    where
        go :: [Int] -> Int -> Int -> [Int]
        go indices currentCapacity k
            | k < 1 || currentCapacity <= 0 = indices
            | otherwise = let (Item weight _) = items ! k  in
                if (currentCapacity < weight) then go indices currentCapacity (k - 1)
                else
                    let previous = if k == 1 then 0 else knapsack ! (currentCapacity, k - 1)
                    in if knapsack ! (currentCapacity, k) /= previous
                        then go (k : indices) (currentCapacity - weight) (k - 1)
                        else go indices currentCapacity (k - 1)
            | otherwise = sort indices


solveKnapsackProblem :: [Item] -> Int -> Int -> [Int]
solveKnapsackProblem items nrItems capacity =
    let xs = listArray (1, nrItems) items
        solve :: Int -> Int -> Int
        solve c jy
            | c == 0 || jy == 0 = 0
            | otherwise = let Item weight value = xs ! jy in
                if c < weight then knapsack ! (c, jy - 1)
                else max (knapsack ! (c - weight, jy - 1) + value) (knapsack ! (c, jy - 1))
        bounds = ((0, 0), (capacity, nrItems))
        knapsack = listArray bounds [solve c k | (c, k) <- range bounds]
    in findItemIndices knapsack (listArray (1, nrItems) items) nrItems capacity


main :: IO()
main = do
    [nrItems, capacity] <- fmap convertToIntList getLine
    items <- readItems nrItems
    let result = solveKnapsackProblem items nrItems capacity
    print $ length result
    putStrLn $ unwords $ map show result

File: Python/22_knapsack.py
import sys

from collections import namedtuple

Item = namedtuple('Item', ['weight', 'value'])


def convert_to_intlist(line):
    return [int(item) for item in line.split()]


def read_items(reader, nr_items):
    items = []
    for _ in range(nr_items):
        items.append(Item(*convert_to_intlist(next(reader))))
    return items


def solve_knapsack_problem(items, nr_items, capacity):
    knapsack = [[-1] * nr_items for _ in range(capacity)]

    def solve(current_capacity, n):
        if current_capacity == 0 or n == 0:
            return 0
        if knapsack[current_capacity-1][n-1] != -1:
           return knapsack[current_capacity-1][n-1]

        weight, value = items[n-1]
        if current_capacity < weight:
            optimal_value = solve(current_capacity, n - 1)
        else:
            optimal_value = max(solve(current_capacity - weight, n - 1) + value, solve(current_capacity, n - 1))
        knapsack[current_capacity-1][n-1] = optimal_value
        return optimal_value

    _ = solve(capacity, nr_items)
    return find_item_indices(knapsack, items, nr_items, capacity)


def find_item_indices(knapsack, items, nr_items, capacity):
    indices = []
    current_capacity = capacity
    n = nr_items
    while n >= 1 and current_capacity > 0:
        weight, _ = items[n-1]
        if current_capacity >= weight:
            previous = 0 if n == 1 else knapsack[current_capacity-1][n-2]
            if knapsack[current_capacity-1][n-1] != previous:
                indices.append(n)
                current_capacity -= weight
        n -= 1
    return sorted(indices)


def main():
    reader = sys.stdin
    nr_items, capacity = convert_to_intlist(next(reader))
    items = read_items(reader, nr_items)
    result = solve_knapsack_problem(items, nr_items, capacity)
    print(len(result))
    print(' '.join([str(elem) for elem in result]))


if __name__ == "__main__":
    main()
